src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java

```diff
@@ -0,0 +1,589 @@
+            // then the parallel stream is slightly faster to get the first result, but for continuous reads it gives us
+            // a higher GC pressure. The sequential stream is slightly slower to get the first result, but it has the
+            // same throughput as the parallel stream, and it gives us less GC pressure.
+            // See the details in the benchmark: https://gist.github.com/Mmuzaf/80c73b7f9441ff21f6d22efe5746541a
+            stream = StreamSupport.stream(data.spliterator(), false)
+                                  .map(row -> makeRow(row, columnFilter))
+                                  .filter(cr -> partitionKey.equals(cr.key.get()))
+                                  .filter(cr -> clusteringFilter.selects(cr.clustering));
+        }
+        else
+        {
+            R row = decorateKeyToRowExtractor.apply(partitionKey);
+            if (row == null)
+                return EmptyIterators.unfilteredPartition(metadata);
+            stream = Stream.of(makeRow(row, columnFilter));
+        }
+
+        // If there are no clustering columns, we've found a unique partition that matches the partition key,
+        // so we can stop the stream without looping through all the rows.
+        if (walker.count(Column.Type.CLUSTERING) == 0)
+            stream.findFirst().ifPresent(cr -> rows.put(cr.clustering, cr.rowSup.get()));
+        else
+            stream.forEach(cr -> rows.put(cr.clustering, cr.rowSup.get()));
+
+        return new SingletonUnfilteredPartitionIterator(new DataRowUnfilteredIterator(partitionKey,
+                                                                                      clusteringFilter,
+                                                                                      columnFilter,
+                                                                                      rows));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public UnfilteredPartitionIterator select(DataRange dataRange, ColumnFilter columnFilter)
+    {
+        return createPartitionIterator(metadata, new AbstractIterator<>()
+        {
+            private final Iterator<? extends UnfilteredRowIterator> partitions = buildDataRangeIterator(dataRange,
+                                                                                                        columnFilter);
+
+            @Override
+            protected UnfilteredRowIterator computeNext()
+            {
+                return partitions.hasNext() ? partitions.next() : endOfData();
+            }
+
+            private Iterator<? extends UnfilteredRowIterator> buildDataRangeIterator(DataRange dataRange,
+                                                                                     ColumnFilter columnFilter)
+            {
+                NavigableMap<DecoratedKey, NavigableMap<Clustering<?>, Row>> partitionMap = new ConcurrentSkipListMap<>(
+                    DecoratedKey.comparator);
```

Comment:
I would violate 120 chars rule if it just reads better, should be on the same line as previous one